src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11800,6 +11869,77 @@ private ExpressionSyntax ParseNewExpression()
+        {
+            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.OpenBracketToken);
+            var openBracket = this.EatToken(SyntaxKind.OpenBracketToken);
+            var list = this.ParseCommaSeparatedSyntaxList(
+                ref openBracket,
+                SyntaxKind.CloseBracketToken,
+                static @this => @this.IsPossibleCollectionElement(),
+                static @this => @this.ParseCollectionElement(),
+                skipBadCollectionElementTokens,
+                allowTrailingSeparator: true,
+                requireOneElement: false,
+                allowSemicolonAsSeparator: false);
+
+            return _syntaxFactory.CollectionCreationExpression(
+                openBracket,
+                list,
+                this.EatToken(SyntaxKind.CloseBracketToken));
+
+            static PostSkipAction skipBadCollectionElementTokens(
+                LanguageParser @this, ref SyntaxToken openBracket, SeparatedSyntaxListBuilder<CollectionElementSyntax> list, SyntaxKind expectedKind, SyntaxKind closeKind)
+            {
+                return @this.SkipBadSeparatedListTokensWithExpectedKind(ref openBracket, list,
+                    static p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleCollectionElement(),
+                    static (p, closeKind) => p.CurrentToken.Kind == closeKind,
+                    expectedKind, closeKind);
+            }
+        }
+
+        private bool IsPossibleCollectionElement()
+        {
+            if (this.IsPossibleExpression())
+                return true;
+
+            // Checking for ':' is for error recovery when someone has a dictionary element and is missing the key part.
+            if (this.CurrentToken.Kind == SyntaxKind.ColonToken)
+                return true;
+
+            // Checking for `keyword:` is for error recovery when typing a dictionary element, but the partial
+            // identifier happens to match a keyword.
+            if (SyntaxFacts.IsReservedKeyword(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken)
+                return true;
+
+            return false;
+        }
+
+        private CollectionElementSyntax ParseCollectionElement()
+        {
+            var dotDotToken = this.TryEatToken(SyntaxKind.DotDotToken);
+            if (dotDotToken != null)
+                return _syntaxFactory.SpreadElement(dotDotToken, this.ParseExpressionCore());
+
+            // Be resilient to `keyword:val` if the user hits that while typing out a full identifier.
+            ExpressionSyntax expression;
+            if (SyntaxFacts.IsReservedKeyword(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken)
+            {
+                var keyword = this.EatTokenWithPrejudice(SyntaxKind.IdentifierToken);
+                var identifier = ConvertToIdentifier(keyword).WithDiagnosticsGreen(keyword.GetDiagnostics());
```

Comment:
>.WithDiagnosticsGreen(keyword.GetDiagnostics()) [](http://example.com/codeflow?start=61&length=47)

Is the `WithDiagnosticsGreen` portion necessary?

Reply:
Yes.  ConvertToIdentifier doesn't preserve diagnostics.

Reply:
Can there be diagnostics? Other places that use `ConvertToIdentifier` don't do this

Reply:
I don't think diagnostics can occur. But if they could, then the `WithDiagnosticsGreen` call should be inside `ConvertToIdentifier` (like we do in `ConvertToKeyword`), not the responsibility of the caller of `ConvertToIdentifier`.

Reply:
> Can there be diagnostics? 

Yes.  EatTokenWithPrejudice adds one.  

> Other places that use ConvertToIdentifier don't do this

They likely aren't using EatTokenWithPrejudice  :)

> But if they could, then the WithDiagnosticsGreen call should be inside ConvertToIdentifier

I can do that and see what happens.